Partial fix for sampling Geometric and Poisson distributions

akash9712 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #13576 

#### Brief description of what is fixed or changed
`sample` method for sampling from Poisson and Geometric method now works.
To do this`_inverse_cdf_expression` was changed in drv.py to isolate and
return the invertible part of the CDF for the expressions.
For sampling from a poisson distribution, the inverse of the CDF is found
via a linear search through the subintervals in the CDF, because solveset is
unable to invert the expression. This, however, works efficiently only for
small values of lamda, which is why this is a partial fix.

Examples:
```
>>> from sympy import *
>>> from sympy.stats import *
>>> X, Y = Geometric('X', S(1)/3), Poisson('Y', 3)
>>> sample(X)
1
>>> sample(Y)
2
```

#### Other comments

akash9712 committed:
Generate Poisson distribution samples

Poisson distribution samples are now generated by generating a random number
in [0, 1] and interating linearly through CDF subintervals, and seeing where
the random number lies. The lower limit of the subinterval from the sample
space is returned.

akash9712 committed:
Changes to `_inverse_cdf_expression` and `sample`

Modified `_inverse_cdf_expression` to isolate the invertible part of the CDF
from the piecewise CDF expression before inversion using solveset.

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into drv_sampling

akash9712 committed:
Partial fix for sampling Geometric and Poisson distributions

`sample` method for sampling from Poisson and Geometric method now works.
To do this`_inverse_cdf_expression` was changed in drv.py to isolate and
return the invertible part of the CDF for the expressions.
For sampling from a poisson distribution, the inverse of the CDF is found
via a linear search through the subintervals in the CDF, because solveset is
unable to invert the expression. This, however, works efficiently only for
small values of lamda, which is why this is a partial fix.

Examples:
```
>>> from sympy import *
>>> from sympy.stats import *
>>> X, Y = Geometric('X', S(1)/3), Poisson('Y', 3)
>>> sample(X)
1
>>> sample(Y)
2
```

normalhuman commented on sympy/stats/drv.py:
```diff
@@ -30,24 +31,29 @@ def check(*args):
+        x = symbols('x', real = True, positive = True,
+         integer = True, cls=Dummy)
+        y, z = symbols('y z', real=True, positive=True, cls=Dummy)
+        cdf_temp = self.cdf(x).subs(x, y)
```
Why not `self.cdf(y)` here, which should be the same? 
And then I don't see x used anymore in this function.  
Also, positive=True implies real=True, so real=True is redundant here.

akash9712 replied:
Here, the original CDF function is piecewise and non-invertible. Using `x` which is specified to be an integer isolates the invertible part of the CDF, but then `solveset` is unable to invert it, because of the constraint. For this reason, `y`, which is declared as an integer substitutes `x` , and using `floor` in `sample` guaranteed that the result is an integer.

normalhuman commented on sympy/stats/drv_types.py:
```diff
@@ -26,6 +28,20 @@ def check(lamda):
+    def sample(self):
+        u = random.uniform(0, 1)
+        inf = self.set.inf
+        interval = Interval(self.cdf(inf), self.cdf(inf + 1))
```
Since you mentioned efficiency, I wonder if the overhead of creating these Interval objects is worth it. It seems your code amounts to the following:
```
u = random.uniform(0, 1)
inf = self.set.inf
while u > self.cdf(inf):
    inf += 1
return inf
```
I think this version puts the meaning across more clearly: we return the smallest number of the form (inf + integer) where cdf is <= u.

akash9712 committed:
Changes to `sample` method for DRVs.

Removed redundant constraints from `sample` method in drv.py, and simplified
the `sample` method in drv_types.py for Poisson RV.

akash9712 commented:
@normalhuman I have made the required changes. Please review.

normalhuman commented on sympy/stats/drv.py:
```diff
@@ -30,24 +31,29 @@ def check(*args):
         Used by sample
         """
-        x, z = symbols('x, z', real=True, positive=True, cls=Dummy)
+        x = symbols('x', positive = True,
```
Python code standard (PEP8): "Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value." https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

normalhuman commented on sympy/stats/drv.py:
```diff
@@ -30,24 +31,29 @@ def check(*args):
+        x = symbols('x', positive = True,
+         integer = True, cls=Dummy)
+        y, z = symbols('y z', positive=True, cls=Dummy)
+        cdf_temp = self.cdf(x).subs(x, y)
```
Still, why `self.cdf(x).subs(x, y)` here instead of `self.cdf(y)`? I see no reason to insert one variable just to immediately replace it with another.

akash9712 replied:
@normalhuman Sorry, you are right. I was substituting the variables to remove the integer constraint from the `cdf` variable, in order to help `solveset`. But `solveset` works fine either way. I have made the necessary changes.

normalhuman commented on sympy/stats/drv.py:
```diff
@@ -30,24 +31,29 @@ def check(*args):
         # Invert CDF
         try:
-            inverse_cdf = list(solveset(self.cdf(x) - z, x))
+            inverse_cdf = solveset(cdf_temp - z, y, domain = S.Reals)
```
Should be no spaces in `domain=S.Reals`

akash9712 committed:
Removed extra whitespaces and extra variable from `sample` in drv.py

normalhuman committed:
while True: is typical in Python

while(True) looks more like JavaScript.

normalhuman commented:
Looks like the best solution with the tools we have. 

Ideally we'd have an inverse for lowergamma, like [Matlab does](https://www.mathworks.com/help/matlab/ref/gammaincinv.html?s_tid=gn_loc_drop) -- then Poisson sampling would be straightforward. But `mpmath` does not implement this, and SymPy by design doesn't deal with numerics on its own, so here we are.
